File: Python/DataStructures/array.py
# Initializing an array
array = [1, 2, 3, 4]
print("Initial array:", array)

# Accessing elements
print("First element:", array[0])  # Accessing the first element
print("Last element:", array[-1])  # Accessing the last element

# Adding elements
array.append(5)  # Adds 5 to the end
print("After append:", array)
array.insert(0, 0)  # Adds 0 to the beginning
print("After insert at beginning:", array)
array.insert(2, 2.5)  # Adds 2.5 at index 2
print("After insert at index 2:", array)

# Removing elements
array.pop()  # Removes the last element
print("After pop:", array)
array.pop(0)  # Removes the first element
print("After removing first element:", array)
del array[1]  # Removes element at index 1
print("After removing element at index 1:", array)

# Modifying elements
array[1] = 10  # Changes the second element to 10
print("After modifying second element:", array)

# Reversing the array
array.reverse()  # Reverses the array
print("After reverse:", array)

# Sorting the array
array.sort()  # Sorts the array alphabetically or numerically
print("After sort:", array)
array.sort(reverse=False)  # Sorts the array numerically in ascending order
print("After numerical sort:", array)

# Finding maximum and minimum
def find_max(arr):
    return max(arr)

def find_min(arr):
    return min(arr)

print("Maximum value:", find_max(array))
print("Minimum value:", find_min(array))

# Finding sum and average
def find_sum(arr):
    return sum(arr)

def find_average(arr):
    return find_sum(arr) / len(arr)

print("Sum of elements:", find_sum(array))
print("Average of elements:", find_average(array))

# Finding frequency of elements
def frequency_count(arr):
    frequency = {}
    for item in arr:
        frequency[item] = frequency.get(item, 0) + 1
    return frequency

print("Frequency count:", frequency_count(array))

# Removing duplicates
def remove_duplicates(arr):
    return list(set(arr))

print("After removing duplicates:", remove_duplicates(array))

# Merging arrays
def merge_arrays(arr1, arr2):
    return arr1 + arr2

array1 = [1, 2, 3]
array2 = [4, 5, 6]
print("Merged array:", merge_arrays(array1, array2))

# Iterating through the array
print("Iterating using for loop:")
for i in range(len(array)):
    print(array[i])

print("Iterating using forEach (list comprehension):")
[print(element) for element in array]

# Using map to create a new array
new_array = list(map(lambda x: x * 2, array))
print("New array with elements multiplied by 2:", new_array)

# Using filter to create a new array
filtered_array = list(filter(lambda x: x > 2, array))
print("Filtered array with elements greater than 2:", filtered_array)

# Using reduce to sum the array elements
from functools import reduce
sum_of_elements = reduce(lambda acc, curr: acc + curr, array)
print("Sum of elements using reduce:", sum_of_elements)

# Finding an element
found = next((x for x in array if x == 2.5), None)
print("Found element (2.5):", found)

# Finding the index of an element
index = array.index(2.5) if 2.5 in array else -1
print("Index of element (2.5):", index)

# Checking if every element passes a test
all_positive = all(x > 0 for x in array)
print("Are all elements positive?:", all_positive)

# Checking if some elements pass a test
some_greater_than_two = any(x > 2 for x in array)
print("Are some elements greater than 2?:", some_greater_than_two)

# Checking if the array includes an element
includes_three = 3 in array
print("Does the array include 3?:", includes_three)

# Using slice to copy part of the array
sliced_array = array[1:3]
print("Sliced array (from index 1 to 3):", sliced_array)

# Joining elements into a string
joined_string = '-'.join(map(str, array))
print("Joined string:", joined_string)

# Finding the length of the array
print("Length of the array:", len(array))

File: Javascript/DataStructures/Array.js
// Initializing an array
let array = [1, 2, 3, 4];
console.log("Initial array:", array);

// Accessing elements
console.log("First element:", array[0]); // Accessing the first element
console.log("Last element:", array[array.length - 1]); // Accessing the last element

// Adding elements
array.push(5); // Adds 5 to the end
console.log("After push:", array);
array.unshift(0); // Adds 0 to the beginning
console.log("After unshift:", array);
array.splice(2, 0, 2.5); // Adds 2.5 at index 2
console.log("After splice (add 2.5 at index 2):", array);

// Removing elements
array.pop(); // Removes the last element
console.log("After pop:", array);
array.shift(); // Removes the first element
console.log("After shift:", array);
array.splice(1, 1); // Removes 1 element at index 1
console.log("After splice (remove at index 1):", array);

// Modifying elements
array[1] = 10; // Changes the second element to 10
console.log("After modifying second element:", array);

// Reversing the array
array.reverse(); // Reverses the array
console.log("After reverse:", array);

// Sorting the array
array.sort(); // Sorts the array alphabetically or numerically
console.log("After sort:", array);
array.sort((a, b) => a - b); // Sorts the array numerically in ascending order
console.log("After numerical sort:", array);

// Finding maximum and minimum
function findMax(arr) {
  return Math.max(...arr);
}
function findMin(arr) {
  return Math.min(...arr);
}
console.log("Maximum value:", findMax(array));
console.log("Minimum value:", findMin(array));

// Finding sum and average
function findSum(arr) {
  return arr.reduce((a, b) => a + b, 0);
}
function findAverage(arr) {
  return findSum(arr) / arr.length;
}
console.log("Sum of elements:", findSum(array));
console.log("Average of elements:", findAverage(array));

// Finding frequency of elements
function frequencyCount(arr) {
  let frequency = {};
  arr.forEach(item => {
    frequency[item] = (frequency[item] || 0) + 1;
  });
  return frequency;
}
console.log("Frequency count:", frequencyCount(array));

// Removing duplicates
function removeDuplicates(arr) {
  return [...new Set(arr)];
}
console.log("After removing duplicates:", removeDuplicates(array));

// Merging arrays
function mergeArrays(arr1, arr2) {
  return arr1.concat(arr2);
}
let array1 = [1, 2, 3];
let array2 = [4, 5, 6];
console.log("Merged array:", mergeArrays(array1, array2));

// Iterating through the array
console.log("Iterating using for loop:");
for (let i = 0; i < array.length; i++) {
  console.log(array[i]);
}
console.log("Iterating using forEach:");
array.forEach(element => console.log(element));

// Using map to create a new array
let newArray = array.map(x => x * 2);
console.log("New array with elements multiplied by 2:", newArray);

// Using filter to create a new array
let filteredArray = array.filter(x => x > 2);
console.log("Filtered array with elements greater than 2:", filteredArray);

// Using reduce to sum the array elements
let sum = array.reduce((acc, curr) => acc + curr, 0);
console.log("Sum of elements using reduce:", sum);

// Finding an element
let found = array.find(x => x === 2.5);
console.log("Found element (2.5):", found);

// Finding the index of an element
let index = array.indexOf(2.5);
console.log("Index of element (2.5):", index);

// Checking if every element passes a test
let allPositive = array.every(x => x > 0);
console.log("Are all elements positive?:", allPositive);

// Checking if some elements pass a test
let someGreaterThanTwo = array.some(x => x > 2);
console.log("Are some elements greater than 2?:", someGreaterThanTwo);

// Checking if the array includes an element
let includesThree = array.includes(3);
console.log("Does the array include 3?:", includesThree);

// Using slice to copy part of the array
let slicedArray = array.slice(1, 3);
console.log("Sliced array (from index 1 to 3):", slicedArray);

// Joining elements into a string
let joinedString = array.join('-');
console.log("Joined string:", joinedString);

// Finding the length of the array
console.log("Length of the array:", array.length);
